### ISSUE TITLE
PNPLoss

### ISSUE THREAD

**Issue by celsofranssa** [KevinMusgrave/pytorch-metric-learning#673]
On the [PNPLoss](https://kevinmusgrave.github.io/pytorch-metric-learning/losses/#pnploss):

```python
losses.PNPLoss(b=2, alpha=1, anneal=0.01, variant="O", **kwargs)
```

I could not find an explanation of the parameters `b`, `alpha`, and `anneal` on the [documentation](https://kevinmusgrave.github.io/pytorch-metric-learning/losses/#pnploss), [code](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/ac607007dc62666f9de850cd5b8e5694ff0da1c2/src/pytorch_metric_learning/losses/pnp_loss.py#L9), related issues (#589) nor even reading the [referenced paper](https://arxiv.org/pdf/2102.04640.pdf).

 It would be great if you could give me some insights about these parameters and how to tune them.

**Comment by celsofranssa** [comment]
And why can't I use another distance with PNPLoss?
```python
AssertionError('PNPLoss requires the distance metric to be CosineSimilarity')
```

**Comment by KevinMusgrave** [comment]
Sorry about that. 

@interestingzhuo we should probably add an explanation of the parameters.

**Comment by interestingzhuo** [comment]
> referenced paper

ok

**Comment by interestingzhuo** [comment]
> Sorry about that.
> 
> @interestingzhuo we should add an explanation of the parameters.

ok

**Comment by interestingzhuo** [comment]
> And why can't I use another distance with PNPLoss?
> 
> ```python
> AssertionError('PNPLoss requires the distance metric to be CosineSimilarity')
> ```

We only implement the cosine similarity, you can also add another distance.

**Comment by interestingzhuo** [comment]
> On the [PNPLoss](https://kevinmusgrave.github.io/pytorch-metric-learning/losses/#pnploss):
> 
> ```python
> losses.PNPLoss(b=2, alpha=1, anneal=0.01, variant="O", **kwargs)
> ```
> 
> I could not find an explanation of the parameters `b`, `alpha`, and `anneal` on the [documentation](https://kevinmusgrave.github.io/pytorch-metric-learning/losses/#pnploss), [code](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/ac607007dc62666f9de850cd5b8e5694ff0da1c2/src/pytorch_metric_learning/losses/pnp_loss.py#L9), related issues (#589) nor even reading the [referenced paper](https://arxiv.org/pdf/2102.04640.pdf).
> 
> It would be great if you could give me some insights about these parameters and how to tune them.

The best setting is b=2, alpha=8, anneal=0.01, variant=""Dq"",

**Comment by KevinMusgrave** [comment]
What is the meaning of each parameter?

**Comment by KevinMusgrave** [comment]
Or what is the effect of each parameter?

**Comment by interestingzhuo** [comment]
> ```python
> variant="O"
> ```

variant  can be  ["Ds", "Dq", "Iu", "Ib", "O"]. It defines the "Variants of PNP". The best is Dq.
alpha: the Hyperparameter of PNP-Dq, The best is 8.
anneal: the best is 0.01, which was studied by smooth AP
b:  The boundary of PNP-Ib

**Comment by interestingzhuo** [comment]
> Sorry about that.
> 
> @interestingzhuo we should add an explanation of the parameters.

How to edit the "https://kevinmusgrave.github.io/pytorch-metric-learning/losses/"?

**Comment by KevinMusgrave** [comment]
Edit this file: https://raw.githubusercontent.com/KevinMusgrave/pytorch-metric-learning/master/docs/losses.md

Search for `## PNPLoss` and add some parameter descriptions like:

```markdown
**Parameters**

- **variant**: ...

- **alpha**: ...
```

**Comment by celsofranssa** [comment]
> > And why can't I use another distance with PNPLoss?
> > ```python
> > AssertionError('PNPLoss requires the distance metric to be CosineSimilarity')
> > ```
> 
> We only implement the cosine similarity, you can also add another distance.

Sure, that's the idea behind the `pytorch-metric-learning`. However, someone has placed an assertion on [PNP loss](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/ac607007dc62666f9de850cd5b8e5694ff0da1c2/src/pytorch_metric_learning/losses/pnp_loss.py) initialization to ensure that the distance is `CosineSimilarity`:

```python
    def __init__(self, b=2, alpha=1, anneal=0.01, variant="O", **kwargs):
        super().__init__(**kwargs)
        c_f.assert_distance_type(self, CosineSimilarity)
        self.b = b
        self.alpha = alpha
        self.anneal = anneal
        self.variant = variant
        if self.variant not in self.VARIANTS:
            raise ValueError(f"variant={variant} but must be one of {self.VARIANTS}")


```

**Comment by celsofranssa** [comment]
> > > And why can't I use another distance with PNPLoss?
> > > ```python
> > > AssertionError('PNPLoss requires the distance metric to be CosineSimilarity')
> > > ```
> > 
> > 
> > We only implement the cosine similarity, you can also add another distance.
> 
> Sure, that's the idea behind the `pytorch-metric-learning`. However, someone has placed an assertion on [PNP loss](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/ac607007dc62666f9de850cd5b8e5694ff0da1c2/src/pytorch_metric_learning/losses/pnp_loss.py) initialization to ensure that the distance is `CosineSimilarity`:
> 
> ```python
>     def __init__(self, b=2, alpha=1, anneal=0.01, variant="O", **kwargs):
>         super().__init__(**kwargs)
>         c_f.assert_distance_type(self, CosineSimilarity)
>         self.b = b
>         self.alpha = alpha
>         self.anneal = anneal
>         self.variant = variant
>         if self.variant not in self.VARIANTS:
>             raise ValueError(f"variant={variant} but must be one of {self.VARIANTS}")
> ```

Hello @interestingzhuo and @KevinMusgrave,
Any workaround here?

**Comment by interestingzhuo** [comment]
> > > > And why can't I use another distance with PNPLoss?
> > > > ```python
> > > > AssertionError('PNPLoss requires the distance metric to be CosineSimilarity')
> > > > ```
> > > 
> > > 
> > > We only implement the cosine similarity, you can also add another distance.
> > 
> > 
> > Sure, that's the idea behind the `pytorch-metric-learning`. However, someone has placed an assertion on [PNP loss](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/ac607007dc62666f9de850cd5b8e5694ff0da1c2/src/pytorch_metric_learning/losses/pnp_loss.py) initialization to ensure that the distance is `CosineSimilarity`:
> > ```python
> >     def __init__(self, b=2, alpha=1, anneal=0.01, variant="O", **kwargs):
> >         super().__init__(**kwargs)
> >         c_f.assert_distance_type(self, CosineSimilarity)
> >         self.b = b
> >         self.alpha = alpha
> >         self.anneal = anneal
> >         self.variant = variant
> >         if self.variant not in self.VARIANTS:
> >             raise ValueError(f"variant={variant} but must be one of {self.VARIANTS}")
> > ```
> 
> Hello @interestingzhuo and @KevinMusgrave, Any workaround here?

We only test the CosineSimilarity for PNP Loss, and other distance will be added soon.

**Comment by KevinMusgrave** [comment]
The documentation has been updated to provide a bit more context: https://kevinmusgrave.github.io/pytorch-metric-learning/losses/#pnploss 

Let us know if you have more questions.
